httpcache.go

```diff
@@ -186,70 +199,67 @@ func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error
 		}
 
 		resp, err = transport.RoundTrip(req)
-		if err == nil && req.Method == "GET" && resp.StatusCode == http.StatusNotModified {
+		if err != nil {
+			if cachedResp != nil && canStaleOnError(req, cachedResp.Header) {
+				// In case of transport failure and stale-if-error activated, returns cached content
+				// when available
+				return cachedResp, nil
+			}
+			t.Cache.Delete(cacheKey)
+			return resp, err
+		}
+
+		if req.Method == "GET" && resp.StatusCode == http.StatusNotModified {
 			// Replace the 304 response with the one from cache, but update with some new headers
 			endToEndHeaders := getEndToEndHeaders(resp.Header)
 			for _, header := range endToEndHeaders {
 				cachedResp.Header[header] = resp.Header[header]
 			}
 			resp = cachedResp
-		} else if (err != nil || (cachedResp != nil && resp.StatusCode >= 500)) &&
-			req.Method == "GET" && canStaleOnError(cachedResp.Header, req.Header) {
+		} else if cachedResp != nil && resp.StatusCode >= 500 && canStaleOnError(req, cachedResp.Header) {
 			// In case of transport failure and stale-if-error activated, returns cached content
 			// when available
 			return cachedResp, nil
-		} else {
-			if err != nil || resp.StatusCode != http.StatusOK {
-				t.Cache.Delete(cacheKey)
-			}
-			if err != nil {
-				return nil, err
-			}
-		}
-	} else {
-		reqCacheControl := parseCacheControl(req.Header)
-		if _, ok := reqCacheControl["only-if-cached"]; ok {
-			resp = newGatewayTimeoutResponse(req)
-		} else {
-			resp, err = transport.RoundTrip(req)
-			if err != nil {
-				return nil, err
-			}
+		} else if resp.StatusCode != http.StatusOK {
+			t.Cache.Delete(cacheKey)
 		}
 	}
 
-	if cacheable && canStore(parseCacheControl(req.Header), parseCacheControl(resp.Header)) {
-		for _, varyKey := range headerAllCommaSepValues(resp.Header, "vary") {
-			varyKey = http.CanonicalHeaderKey(varyKey)
-			fakeHeader := "X-Varied-" + varyKey
-			reqValue := req.Header.Get(varyKey)
-			if reqValue != "" {
-				resp.Header.Set(fakeHeader, reqValue)
-			}
+	if !canStore(parseCacheControl(req.Header), parseCacheControl(resp.Header)) {
+		t.Cache.Delete(cacheKey)
+		return resp, nil
```

Comment:
since we are using named returns would just `return` be appropiate?

Reply:
> since we are using named returns would just `return` be appropiate?

Yeah, this is a style call that could go either way.  If this project doesn't have it's own opinions, [Go's suggestions][1] include:

> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values.

so at least by that guideline this would fall into the "explicit returns" category.

[1]: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters